wled00/espnow_broadcast.cpp

```diff
@@ -0,0 +1,249 @@
+
+#ifndef WLED_DISABLE_ESPNOW_NEW
+#include <espnow_broadcast.h>
+
+#if defined ESP32
+#include <WiFi.h>
+#include <esp_wifi.h>
+#include <esp_now.h>
+
+#elif defined ESP8266
+#include <ESP8266WiFi.h>
+#define WIFI_MODE_STA WIFI_STA 
+#else
+#error "Unsupported platform"
+#endif //ESP32
+
+#define BROADCAST_ADDR_ARRAY_INITIALIZER {0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF}
+#define WLED_ESPNOW_WIFI_CHANNEL 1
+
+typedef struct {
+    uint8_t mac[6];
+    uint8_t len;
+    uint8_t data[WLED_ESPNOW_MAX_MESSAGE_LENGTH];
+} QueuedNetworkMessage;
+static_assert(WLED_ESPNOW_MAX_MESSAGE_LENGTH <= ESP_NOW_MAX_DATA_LEN, "WLED_ESPNOW_MAX_MESSAGE_LENGTH must be <= 250 bytes");
+
+ESPNOWBroadcast espnowBroadcast;
+
+ESPNOWBroadcast::ESPNOWBroadcast() {    
+    _networkRxMessages = xRingbufferCreateNoSplit(sizeof(QueuedNetworkMessage), WLED_ESPNOW_MAX_QUEUED_MESSAGES);
+}
+
+bool ESPNOWBroadcast::setup() {
+
+    auto err = esp_event_loop_create_default();
+    if ( ERR_OK != err && ESP_ERR_INVALID_STATE != err ) {
+        return false;
+    }
+    err = esp_event_handler_instance_register(WIFI_EVENT, WIFI_EVENT_STA_START, onWiFiEvent, NULL, NULL);
+    if ( ERR_OK != err ) {
+        return false;
+    }
+    err = esp_event_handler_instance_register(WIFI_EVENT, WIFI_EVENT_STA_STOP, onWiFiEvent, NULL, NULL);
+    if ( ERR_OK != err ) {
+        return false;
+    }
+    err = esp_event_handler_instance_register(WIFI_EVENT, WIFI_EVENT_AP_START, onWiFiEvent, NULL, NULL);
+    if ( ERR_OK != err ) {
+        return false;
+    }
+
+    return setupWiFi();
+}
+
+bool ESPNOWBroadcast::setupWiFi() {
+    Serial.println("ESPNOWBroadcast::setupWiFi()");
+
+    _state.exchange(STOPPED);
+
+    // To enable ESPNow, we need to be in WIFI_STA mode
+    if ( !WiFi.mode(WIFI_STA) ) {
+        Serial.println("WiFi.mode() failed");
+        return false;
+    }
+    // and not have the WiFi connect
+    // Calling discount with tigger an async Wifi Event
+    if ( !WiFi.disconnect(false, true) ) {
+        Serial.println("WiFi.disconnect() failed");
+        return false;
+    }
+
+    return true;
+}
+
+
+void ESPNOWBroadcast::loop(size_t maxMessagesToProcess /*= 1*/) {
+    switch (espnowBroadcast._state.load()) {
+        case ESPNOWBroadcast::STARTING:
+            // if WiFI is in starting state, actually stat ESPNow from our main task thread.
+            start();
+            break;
+        case ESPNOWBroadcast::STARTED: {
+            auto ndx = maxMessagesToProcess;
+            while(ndx-- > 0) {
+                size_t size = 0;
+                auto *msg = (QueuedNetworkMessage*)xRingbufferReceive(_networkRxMessages, &size, 0);
+                if (msg) {
+                    if(size == sizeof(*msg)) {
+                        auto callback = _rxCallback;
+                        while( *callback ) {
+                            (*callback)(msg->mac, msg->data, msg->len);

```

Comment:
Since we're hooking this into WLED's core code
Let's add functionality to match base WLED's use of espnow: to filter by mac and not even call the callbacks if the mac isn't the one we're listening to.
And `listener == FFFFFFFFFFFF` is a way to say that you're listening to anyone.  (It could even be the default value.)

Reply:
future enhancement. and I'm not sure it should be in the ESPNOWBroadcast class vs its consumer